### PR TITLE
Add _format_download_uri_for_extension method

### PR DESCRIPTION
**Pull request by c-w**
In https://github.com/c-w/gutenberg/issues/105 we saw that there should be a method to leverage the URI building functionality of the Gutenberg library for non-text resources. This pull request implements `_format_download_uri_for_extension` which is a refactoring of our existing URI formatting logic that can take an arbitrary file extension like `-pdf.pdf` (as for example returned by the `formaturi` metadata call) and format the appropriate URL.

This is not an official API (whence the underscore prefix) so no strict requirement to always maintain this function, but it's useful functionality to expose nevertheless given that we already have it in the code base.

When working on this change, I also noticed that we weren't correctly handling the case where the mirror root URL was changed in a single session so I also added a fix for that.

When verifying the refactor, I saw some odd behavior when running the tests locally, so this pull request also improves the developer experience for the tests:

- We're now using the [responses](https://pypi.org/project/responses/) library to stub out the `requests.head` calls instead of doing it manually.
- We're now running the tests in a Docker container to ensure that the environment is always known. E.g. for Python 2, BSD-DB should not be installed but it should be installed for Python 3.
- We're now using `python setup.py install` to install the dependencies for the various runtimes since the setup file already knows how to load the correct pip dependencies for Python 2 vs Python 3. Previously on Travis we were always installing the Python 3 dependencies even when running the tests for Python 2 which could lead to odd behavior.

### CODE REVIEW

**Comment by hugovk** [comment]
`python setup.py install` to install the correct dependencies is a good idea. 

Would it be better to instead call `pip install .` for the reasons given here?

http://python3statement.org/practicalities/#unit-testing-and-documentation

**Comment by c-w** [comment]
Interesting, didn't know about that. Fixed in ee601ae.

File: gutenberg/acquire/text.py [hunk 1]
```diff
@@ -7,17 +7,23 @@
 import os
 from contextlib import closing
 
+try:
+    from functools import lru_cache
+except ImportError:
+    from functools32 import lru_cache
```
Comment (hugovk) [inline gutenberg/acquire/text.py]: Is this due to a Python 2/3 difference?

If so, commenting which is which will make dropping Python 2 easier in the near future.

Comment (c-w) [reply gutenberg/acquire/text.py]: Yes, `functools.lru_cache` was only added in Python 3. the functools32 package is a backport for Python 2.7 I'll add a comment.

Comment (c-w) [reply gutenberg/acquire/text.py]: Done in e49912f.

**Pull request merged**